File: .cache/claude-usage-monitor.py
#!/usr/bin/env python3
"""
Claude API Usage Monitor
Fetches usage data from Anthropic Console and tracks costs with alerts.
"""

import json
import os
import sys
from datetime import datetime
from pathlib import Path
import subprocess
import requests

# Configuration
CACHE_FILE = Path.home() / ".openclaw/workspace/.cache/claude-usage.json"
WEBHOOK_URL = os.getenv("WEBHOOK_URL", "")

# Pricing rates (2026)
INPUT_RATE = 0.4 / 1_000_000  # $0.4 per 1M tokens
OUTPUT_RATE = 1.2 / 1_000_000  # $1.2 per 1M tokens

# Budgets
DAILY_BUDGET = 5.00
MONTHLY_BUDGET = 155.00
DAILY_ALERT_THRESHOLD = 3.75  # 75%
MONTHLY_ALERT_THRESHOLD = 116.25  # 75%


def calculate_cost(input_tokens: int, output_tokens: int) -> float:
    """Calculate cost from token counts."""
    return (input_tokens * INPUT_RATE) + (output_tokens * OUTPUT_RATE)


def fetch_from_console_browser() -> dict:
    """
    Fetch usage data using browser automation.
    Requires chromium and playwright or similar.
    """
    try:
        result = subprocess.run(
            [
                "node",
                "-e",
                """
const playwright = require('playwright');
(async () => {
    const browser = await playwright.chromium.launch();
    const page = await browser.newPage();
    await page.goto('https://console.anthropic.com/account/usage');
    
    // Wait for data to load
    await page.waitForLoadState('networkidle');
    
    // Try to extract usage data from the page
    const data = await page.evaluate(() => {
        // Look for usage stats in the DOM
        const tokens = document.querySelector('[data-testid="usage-tokens"]')?.textContent || null;
        const cost = document.querySelector('[data-testid="usage-cost"]')?.textContent || null;
        return { tokens, cost };
    });
    
    console.log(JSON.stringify(data));
    await browser.close();
})();
                """,
            ],
            capture_output=True,
            text=True,
            timeout=30,
        )

        if result.returncode == 0:
            return json.loads(result.stdout)
        else:
            print(f"Browser automation failed: {result.stderr}", file=sys.stderr)
            return None
    except Exception as e:
        print(f"Could not use browser automation: {e}", file=sys.stderr)
        return None


def log_usage(
    input_tokens_today: int,
    output_tokens_today: int,
    input_tokens_month: int,
    output_tokens_month: int,
):
    """Log usage data and check thresholds."""

    timestamp = datetime.utcnow().isoformat() + "Z"
    tokens_today = input_tokens_today + output_tokens_today
    tokens_month = input_tokens_month + output_tokens_month

    cost_today = calculate_cost(input_tokens_today, output_tokens_today)
    cost_month = calculate_cost(input_tokens_month, output_tokens_month)

    percent_daily = (cost_today / DAILY_BUDGET) * 100
    percent_monthly = (cost_month / MONTHLY_BUDGET) * 100

    # Determine status
    status = "ok"
    alert_triggered = False

    if cost_today > DAILY_ALERT_THRESHOLD:
        status = "warning_daily"
        alert_triggered = True
        print(f"⚠️  Daily budget alert: ${cost_today:.4f} / ${DAILY_BUDGET:.2f}")

    if cost_month > MONTHLY_ALERT_THRESHOLD:
        status = "warning_monthly"
        alert_triggered = True
        print(f"⚠️  Monthly budget alert: ${cost_month:.4f} / ${MONTHLY_BUDGET:.2f}")

    # Build log entry
    log_entry = {
        "timestamp": timestamp,
        "tokens_today": tokens_today,
        "input_tokens_today": input_tokens_today,
        "output_tokens_today": output_tokens_today,
        "cost_today": round(cost_today, 4),
        "tokens_month": tokens_month,
        "input_tokens_month": input_tokens_month,
        "output_tokens_month": output_tokens_month,
        "cost_month": round(cost_month, 4),
        "budget_daily": DAILY_BUDGET,
        "budget_monthly": MONTHLY_BUDGET,
        "percent_daily": round(percent_daily, 1),
        "percent_monthly": round(percent_monthly, 1),
        "status": status,
    }

    # Write to cache
    CACHE_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(CACHE_FILE, "w") as f:
        json.dump(log_entry, f, indent=2)

    print(f"✅ Logged to {CACHE_FILE}")
    print(json.dumps(log_entry, indent=2))

    # Send webhook if needed
    if alert_triggered and WEBHOOK_URL:
        print(f"🔔 Triggering webhook alert...")
        try:
            response = requests.post(
                WEBHOOK_URL, json=log_entry, timeout=10
            )
            response.raise_for_status()
            print("✅ Webhook sent")
        except Exception as e:
            print(f"❌ Webhook failed: {e}")

    return log_entry


def main():
    """Main entry point."""
    print("📊 Claude API Usage Monitor")
    print(f"⏰ {datetime.utcnow().isoformat()}Z")
    print()

    # Try to get data from environment variables first
    usage_daily = os.getenv("USAGE_DATA_DAILY", "")
    usage_month = os.getenv("USAGE_DATA_MONTH", "")

    if usage_daily and usage_month:
        # Parse: "input=123456 output=789012"
        daily_parts = dict(
            part.split("=") for part in usage_daily.split()
        )
        month_parts = dict(
            part.split("=") for part in usage_month.split()
        )

        input_today = int(daily_parts.get("input", 0))
        output_today = int(daily_parts.get("output", 0))
        input_month = int(month_parts.get("input", 0))
        output_month = int(month_parts.get("output", 0))

        print("✅ Using environment data")
        log_usage(input_today, output_today, input_month, output_month)
    else:
        print(
            "ℹ️  No environment data. To use this script, set:"
        )
        print(
            "  export USAGE_DATA_DAILY='input=123456 output=789012'"
        )
        print(
            "  export USAGE_DATA_MONTH='input=1234567 output=7890123'"
        )
        print()
        print("Or configure browser automation for automatic fetching.")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: .cache/claude-usage-monitor.js
#!/usr/bin/env node

/**
 * Claude API Usage Monitor
 * Tracks daily/monthly costs and alerts when thresholds are exceeded
 * 
 * Haiku rates (April 2026):
 * - Input:  $0.40 per 1M tokens
 * - Output: $1.20 per 1M tokens
 */

const fs = require('fs');
const path = require('path');
const https = require('https');

// Config
const CACHE_DIR = path.join(__dirname);
const USAGE_FILE = path.join(CACHE_DIR, 'claude-usage.json');
const WEBHOOK_URL = process.env.WEBHOOK_MONITOR || null;

const HAIKU_RATES = {
  input: 0.40 / 1_000_000,   // $0.40 per 1M
  output: 1.20 / 1_000_000   // $1.20 per 1M
};

const BUDGETS = {
  daily: 5.00,
  monthly: 155.00,
  alert_threshold: 0.75 // Alert at 75% of budget
};

/**
 * Parse usage data from Anthropic console
 * Expected input format:
 * {
 *   tokens_today: { input: 1000, output: 5000 },
 *   tokens_month: { input: 50000, output: 250000 }
 * }
 */
function calculateCosts(tokensInput, tokensOutput) {
  return {
    input_cost: tokensInput * HAIKU_RATES.input,
    output_cost: tokensOutput * HAIKU_RATES.output,
    total_cost: (tokensInput * HAIKU_RATES.input) + (tokensOutput * HAIKU_RATES.output)
  };
}

function updateUsage(todayData, monthData) {
  const timestamp = new Date().toISOString();
  
  const costToday = calculateCosts(
    todayData.input || 0,
    todayData.output || 0
  );
  
  const costMonth = calculateCosts(
    monthData.input || 0,
    monthData.output || 0
  );

  const dailyPercent = (costToday.total_cost / BUDGETS.daily) * 100;
  const monthlyPercent = (costMonth.total_cost / BUDGETS.monthly) * 100;

  let status = 'OK';
  let alert = false;

  if (costToday.total_cost > (BUDGETS.daily * BUDGETS.alert_threshold)) {
    status = 'ALERT_DAILY';
    alert = true;
  } else if (costMonth.total_cost > (BUDGETS.monthly * BUDGETS.alert_threshold)) {
    status = 'ALERT_MONTHLY';
    alert = true;
  }

  const record = {
    timestamp,
    tokens_today: { input: todayData.input || 0, output: todayData.output || 0 },
    cost_today: parseFloat(costToday.total_cost.toFixed(4)),
    percent_daily: parseFloat(dailyPercent.toFixed(2)),
    tokens_month: { input: monthData.input || 0, output: monthData.output || 0 },
    cost_month: parseFloat(costMonth.total_cost.toFixed(4)),
    percent_monthly: parseFloat(monthlyPercent.toFixed(2)),
    budget_daily: BUDGETS.daily,
    budget_monthly: BUDGETS.monthly,
    alert_threshold: BUDGETS.alert_threshold * 100,
    status
  };

  // Write log
  fs.writeFileSync(USAGE_FILE, JSON.stringify(record, null, 2));
  console.log(`[${timestamp}] Usage logged: ${status}`);
  console.log(`  Daily: $${record.cost_today.toFixed(2)} / $${BUDGETS.daily} (${record.percent_daily}%)`);
  console.log(`  Monthly: $${record.cost_month.toFixed(2)} / $${BUDGETS.monthly} (${record.percent_monthly}%)`);

  // Trigger webhook if alert
  if (alert && WEBHOOK_URL) {
    sendWebhookAlert(record);
  }

  return record;
}

function sendWebhookAlert(record) {
  const payload = JSON.stringify({
    type: 'claude-api-budget-alert',
    severity: record.status === 'ALERT_DAILY' ? 'critical' : 'warning',
    message: `Claude API usage at ${record.percent_daily}% daily / ${record.percent_monthly}% monthly`,
    data: record,
    timestamp: new Date().toISOString()
  });

  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': payload.length
    }
  };

  const req = https.request(WEBHOOK_URL, options, (res) => {
    console.log(`[WEBHOOK] Status: ${res.statusCode}`);
  });

  req.on('error', (err) => {
    console.error(`[WEBHOOK] Error: ${err.message}`);
  });

  req.write(payload);
  req.end();
}

/**
 * Main entry point
 * 
 * Usage from cron:
 * node .cache/claude-usage-monitor.js --today-input 1000 --today-output 5000 --month-input 50000 --month-output 250000
 */
function main() {
  const args = process.argv.slice(2);
  
  let todayInput = 0, todayOutput = 0, monthInput = 0, monthOutput = 0;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--today-input') todayInput = parseInt(args[++i], 10);
    if (args[i] === '--today-output') todayOutput = parseInt(args[++i], 10);
    if (args[i] === '--month-input') monthInput = parseInt(args[++i], 10);
    if (args[i] === '--month-output') monthOutput = parseInt(args[++i], 10);
  }

  if (!todayInput && !todayOutput && !monthInput && !monthOutput) {
    console.error('No usage data provided. Usage:');
    console.error('  node claude-usage-monitor.js --today-input N --today-output N --month-input N --month-output N');
    process.exit(1);
  }

  updateUsage(
    { input: todayInput, output: todayOutput },
    { input: monthInput, output: monthOutput }
  );
}

if (require.main === module) {
  main();
}

module.exports = { updateUsage, calculateCosts, HAIKU_RATES, BUDGETS };
